Revbump to SoCo 0.13, add support for Night Sound and Speech Enhancement

rbdixon opened the pull request:
Ready to pull.

## Description:

Sonos Playbar and Playbase devices support Night Sound and Speech Enhancement
effects when playing from sources such as a TV. Adds a new service "sonos_set_option"
whichs accepts boolean options to control these audio features.

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

homeassistant commented:
Hi @rbdixon, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10765).

Once you do that we will be able to review and accept this pull request.

Thanks!

rbdixon committed:
Revbump to SoCo 0.13 and add support for Night Sound and Speech Enhancement.

Sonos Playbar and Playbase devices support Night Sound and Speech Enhancement
effects when playing from sources such as a TV. Adds a new service "sonos_set_option"
whichs accepts boolean options to control these audio features.

rbdixon commented:
I've added docs [PR](https://github.com/home-assistant/home-assistant.github.io/pull/4107) and I think this is ready to merge. I cobbled together a simple test based on the patterns I saw for testing the other features. Local testing with my speakers works fine, too, of course.

amelchio commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -834,6 +858,22 @@ def media_title(self):
+    @property
+    def night_sound(self):
+        """Get status of Night Sound."""
+        if self._coordinator:
```
You defer to the coordinator. Are these options really per-group (like a playlist), not per-speaker (like volume)?

rbdixon replied:
I did some research using two speakers (Kitchen, Den) of my multi-unit setup. Den is a Playbar that supports the night_sound (NS) and speech_enhance (SE) features when playing from the TV source.

1. When I build a group designating Kitchen as the coordinator and then switch to the TV source the Den then becomes the coordinator.
2. When playing TV in a group the NS and SE effects are audible on the Den but not the Kitchen.
3. SoCo will not send NS and SE commands to a non-Playbar speaker.

Based on #2 I agree with you that this is a per-speaker feature. This was masked by the Sonos behavior described in #1.

I'll push a new version that does not defer to the coordinator momentarily. It will not include a check to ensure that the speaker is the coordinator but given #3 this should not cause an error.

Sound good? Thank you for the close review.

rbdixon replied:
I pushed an updated version per your comments. I also rebased to the current dev branch.

amelchio replied:
I like your changes though if the Playbar always ends up being the coordinator, I guess it doesn't matter much.

amelchio reviewed (approved):
Looks good 👍. I hope you will help move this to events when/if #10419 gets merged.
